### PR TITLE
add documentation for new operator interface in operator_util.h

### PR DESCRIPTION
**Pull request by precedenceguo**
Suggestions welcome

### CODE REVIEW

File: docs/system/operator_util.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+that functions can be separate out by the number of operands. Derivatives are a bit more 
+complex. Whether output value, input data or neither are needed alongside head gradient is 
+crucial to construct a dependency graph. Gradient functions in the unified API is thus 
+differentiated through the types of operands it takes for calculation.
+
+Before we continue on the operator interface, it is recommend to take a look at the [mshadow
+library guide](https://github.com/dmlc/mshadow/tree/master/guide) since actual calculations 
+will be done in `mshadow::TBlob` structure.
+
+# Unified Operator API
```
Comment (tqchen) [inline docs/system/operator_util.md]: only use first level header at title

Comment (tqchen) [inline docs/system/operator_util.md]: e.g. use ```##```

**Comment by tqchen** [comment]
thanks for contribution! The guide looks good.

Can you directly add an example code to the description? So the user who want to  learn how to do so can look at the example besides the instructions. I also made a minor comment on header level

**Comment by tqchen** [comment]
Thanks! This looks great, let us add link to it in the system/index.md as well as how_to/new_op.md and we can merge it in

**Comment by precedenceguo** [comment]
Let's try if travis works after rebase to the latest master.

**Comment by piiswrong** [comment]
lets also put this under what's new. BTW do we have a example using keyword argument? like sum(axis=1)

**Comment by precedenceguo** [comment]
Question: loss in mxnet seems to have a convention that only the gradient is given. e.g. we don't have mean cross entropy loss value from SoftmaxOutput, or mean squared error from LinearRegressionOutput. They seem to be implemented in mxnet.metric. In the example of smooth l1 loss, I did calculate the loss itself as well as the gradient. Do we need to factor smooth l1 loss to  that convention? And should it be documented in the guide?

**Comment by piiswrong** [comment]
We should have both: direct output for simple loss and sumed loss value for composition.
This should be carefully documented though

**Comment by precedenceguo** [comment]
That's right. We can composite linear regression from mx.sym.square(in - out). Then it is a direct output for simple loss. With modifications to `mx.sym.MakeLoss` we can sum loss value and output directly. Existing `mx.sym.LinearRegressionOutput` can cope with `mx.metric` for summed loss value. I think it would be appropriate to leave this documentation in the operator description instead of a developer guide. I will modify `MakeLoss` for this behavior in next pr.

**Comment by precedenceguo** [comment]
As for the keyword arguments, I am so sure how to use that appropriately and need an example myself. And where is `what's new`?
Anyway I think this doc is ready for now. Maybe we should fix minor details later.

File: src/operator/smooth_l1_unary-inl.h [hunk 2]
```diff
@@ -104,7 +104,7 @@ void SmoothL1BackwardUseIn_(const OutputGrad& out_grad,
 }
 
 MXNET_REGISTER_SIMPLE_OP(smooth_l1, XPU)
-.set_function(XPU::kDevMask, SmoothL1Forward_<XPU>, kInplaceInOut)
+.set_function(XPU::kDevMask, SmoothL1Forward_<XPU>, kNoInplace)
```
Comment (piiswrong) [inline src/operator/smooth_l1_unary-inl.h]: I just noticed this. Is this correct? Was the previous version buggy?

Comment (precedenceguo) [inline src/operator/smooth_l1_unary-inl.h]: We have DeclareBackwardDependency to ensure the in_data does not get
replaced. Previous version worked correctly thanks to this. I would say
Runtime Engine is quite robust. Let us change it now.
